### ISSUE TITLE
fabric-net-svpc-access uses authoritative iam_binding

### ISSUE THREAD

**Issue by ideasculptor** [terraform-google-modules/terraform-google-network#98]
The module surely needs to use iam_member instead of iam_binding, or provide the same 'additive' option that the iam modules provide - easily done by simply using the subnets_iam module (https://github.com/terraform-google-modules/terraform-google-iam/tree/master/modules/subnets_iam )

If you use the module as it is, you are pretty much guaranteed to break stuff

**Comment by ludoo** [comment]
I'm not convinced we should change it, as this is the right place to handle subnet IAM bindings, and they then need to be authoritative to ovverride any potential drift. Keep in mind this is a "fabric" module, and while it can definitely be used standalone, it's meant to be composed with others and "own" its tiny piece of infra.

**Comment by ludoo** [comment]
If you feel strongly about this, please bring actual use cases, as we've been using this module in actual code and it works well for our use case.

**Comment by ideasculptor** [comment]
My use case is pretty simple - I have more than one team responsible for networks.  Each team wants to be able to manage their subnets, routes, and firewall rules separately.  The team responsible for GKE subnets don't want to have to deal with subnets set up for GCE resources that are independent of the kubernetes networks - communication between k8s and other GCE resources is all pubsub so direct communication isn't necessary and defense-in-depth dictates keeping those subnets separate.  By defining them in separate tempaltes, there is absolutely no risk that one might be misconfigured to grant access to the other.

Additionally, I have a service project for GKE resources, and a host project for each of dev, stage, and prod environments.  A separate service project would be used for the admin and GCE resources, which also need to run in all 3 environment networks.  So the ability to link a single service account to multiple host projects is fairly crucial, and defning the subnets in separate templates means that using remote state to access values for one set of subnets carries no risk that someone might use a subnet from the other template by mistake.

My infrastructure code defines the host project and the host network in one terraform template.  There's a GKE template that sets up the GKE subnets and secondary ranges, and another which creates an administrative subnet for bastions.  A third sets up subnets for GCE resources that support internal services.

This isn't the production code, but a testbed that prototypes the structures:

https://github.com/ideasculptor/gcloud-reference-arch/tree/master/Reference%20Infrastructure/dev/us-central1

If you look inside the 'dev' subdir, which contains resources in the dev environment (hosts the dev network), you can see public_subnets and backend_subnets defined separately.

If you look inside the shared_svc subdir, which contains resources assigned to a service project, you can see the bastion and gke templates are separated, and each only needs to access the remote_state for the subnets that they access.

You'll find the actual terraform code, separated from the config, in https://github.com/ideasculptor/gcloud-templates - the URLs are in the terragrunt.hcl files, along with variable values.  Other variables are defined in .yaml files in the directory tree, to allow values to be inherited just by placing a terragrunt.hcl file in a particular location in the directory tree.

**Comment by ludoo** [comment]
I got more than I bargained for. :) Let me digest this and get back to you.

**Comment by ideasculptor** [comment]
I rewrote it. Probably made it longer, but pointed my explanation more directly at your question

**Comment by ideasculptor** [comment]
The long and the short of it is that an infrastructure for an enterprise with thousands of developers working on hundreds of services and applications is never going to want to define all networking components in a single module call. Arguably, even a company with 100 developers would want more granularity than that, in all likelihood.  And I speak from experience on that point.

**Comment by ludoo** [comment]
Ok it's clear. All my customers tend to define networking in a single point, subject to the same review process, regardless of which team actually manages individual resources (applied to your case, the GKE people would manage their subnet configurations, but they would reside in the same root module and go through the same approval process).

TL;DR: your use case never came to mind, but is of course perfectly valid.

I would still like to preserve a way of using authoritative bindings, so would an extra `use_authoritative_iam_bindings` bool variable in the module with a `true` default, used to decide which type of bindings to apply, and of course an extra non-authoritative resource, cover your use case?

And thanks a lot for bringing this to our attention, of course. Super useful and interesting.

**Comment by ideasculptor** [comment]
I would solve it by using the fabric iam modules, which already allow you to choose between additive and authoritative modes, and there's already a module for subnet iam. You can simply switch to calling that module and pass the type flag through, and it all should 'just work'

**Comment by ideasculptor** [comment]
https://github.com/terraform-google-modules/terraform-google-iam/tree/master/modules/subnets_iam

**Comment by ludoo** [comment]
In reply to your last comment, I'd argue that a large infra is composed of many environments, and many different networks, so you'd have many single touchpoints. But that's immaterial, let's focus on how to address your perfectly valid use case.

I don't get the reference to the IAM module. If you prefer using that, then just create your own module to define project association to the SVPC, and handle the rest in the IAM module. Or am I missing something?

**Comment by ludoo** [comment]
To put my comment above in the proper context: this submodule (and the overall fabric approach) is not to be the single solution encompassing all possible approaches to handling SVPC access, but a convenience module to use as a sensible shortcut while prototyping, and a starting point to fork and evolve when adaptation to specific use cases is needed.

So if the IAM module solves 90% of your needs, I don't think we should change anything here.

**Comment by ideasculptor** [comment]
It just seems like a very fast way to replace that single resource with a single call to an already existing (and tested, presumably) module, in order to get exactly the functionality desired - a way to assign the basic IAM roles to the various service accounts and identities that need them in order to set up access to a host network.  It is just code re-use, nothing more.  No need to re-implement and re-test it in the fabric-ne-svpc-access module when you can just call the other module, insyead.

The only other way I could think of to really solve it would be to use separate templates to define the subnet maps and lists and store them in remote state, then use the module in authoritative mode to pull them all into the single module call - but writing unit and integration tests for that would be such a headache.  In the end, using an additive mode makes testing massively easier.

**Comment by ideasculptor** [comment]
And why make public modules if they aren't intended to be reused?  If they are only there to serve as a copy/paste template or as the root of a fork that never merges back, you lose pretty much all of the benefit of open-source and community effort. That's what causes open source efforts to wallow and never gain widespread adoption. 

It seems far superior, to me, to construct modules that are sufficiently general purpose that they can be used without modification, so that users automatically get the benefit of future bug fixes, future features, and the efforts of all the other users.

**Comment by ludoo** [comment]
Our approach (which has been tested in several very large migrations for 1.5 years), is to minimize complexity and use simple modules as a starting point, and specialize them for the specific use case. Which also works very well for the prototyping use case, which is something my team does a lot (how would we be able to automate testing of new features and configurations otherwise?).

Just look at the host of issues in the project-factory and networking modules lately, where complexity is starting to rear its head. Again, if 90% of your use case is solved by the IAm module, I don't think we need to change anything here. If on the other hand you like having everything related to SVPC access encapsulated in a single module, and your use case mandates non-authoritative IAm resources, let's discuss how to implement that here.

**Comment by morgante** [comment]
> And why make public modules if they aren't intended to be reused?

I want to be clear that all our main modules are meant to be reused. CFT Fabric (which ludoo@ maintains) offers an opinionated subset which don't necessarily follow the same philosophy.

**Comment by ludoo** [comment]
I'm going to close this, feel free to reopen if you need us to implement simple non-authoritative bindings for subnets alongside the authoritative ones.
